### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Bicategory): Grothendieck construction for a pseudofunctor

### PR DESCRIPTION
**Pull request by callesonne**
In this PR we add the Grothendieck construction for pseudofunctors.

---

- [x] depends on: #16119

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+structure Pseudofunctor.Grothendieck (F : Pseudofunctor (LocallyDiscrete 𝒮ᵒᵖ) Cat.{v₂, u₂}) where
+  /-- The underlying object in the base category. -/
+  base : 𝒮
+  /-- The object in the fiber of the base object. -/
+  fiber : F.obj ⟨op base⟩
+
+/-- Notation for the Grothendieck category associated to a pseudofunctor `F`. -/
+prefix:75 "∫ " => Pseudofunctor.Grothendieck
+
+namespace Pseudofunctor.Grothendieck
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:55]: The notation "∫" is used in other parts of mathlib. The following should make "open Pseudofunctor.Grothendieck" necessary to get this notation:
```suggestion
namespace Pseudofunctor.Grothendieck

scoped prefix:75 "∫ " => Pseudofunctor.Grothendieck
```

Comment (callesonne) [reply Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:55]: Thanks!! I'm somewhat new to defining notation :)

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  base : X.base ⟶ Y.base
+  /-- The morphism in the fiber over the domain. -/
+  fiber : X.fiber ⟶ (F.map base.op.toLoc).obj Y.fiber
+
+@[simps!]
+instance categoryStruct : CategoryStruct (∫ F) where
+  Hom X Y := Hom X Y
+  id X := {
+    base := 𝟙 X.base
+    fiber := (F.mapId ⟨op X.base⟩).inv.app X.fiber}
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:65]:
```suggestion
    fiber := (F.mapId ⟨op X.base⟩).inv.app X.fiber }
```

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    · simp
+    · dsimp
+      slice_lhs 3 4 => rw [← (F.mapComp g.base.op.toLoc f.base.op.toLoc).inv.naturality h.fiber]
+      simp [F.mapComp_assoc_right_inv_app, Strict.associator_eqToIso]
+
+
+/-- The projection `∫ F ⥤ 𝒮` given by projecting both objects and homs to the first
+factor. -/
+@[simps]
+def forget (F : Pseudofunctor (LocallyDiscrete 𝒮ᵒᵖ) Cat.{v₂, u₂}) : ∫ F ⥤ 𝒮 where
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:118]: This should be fine if you adding `variable (F)` before the declaration:
```suggestion
def forget : ∫ F ⥤ 𝒮 where
```

**Comment by joelriou** [comment]
Otherwise, it looks good to me!

**Comment by callesonne** [comment]
Thanks a lot for the review @joelriou !

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+* Morphisms: morphisms `(R, b) ⟶ (S, a)` are defined as pairs `(f, h)` where `f : R ⟶ S` is a
+  morphism in `𝒮` and `h : b ⟶ F(f)(a)`
+
+The projection functor `∫ F ⥤ 𝒮` is then given by projecting to the first factors, i.e.
+* On objects, it sends `(S, a)` to `S`
+* On morphisms, it sends `(f, h)` to `f`
+
+## References
+[Vistoli2008] "Notes on Grothendieck Topologies, Fibered Categories and Descent Theory" by
+Angelo Vistoli
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:27]: You should add a bibtex format entry in the file `docs/references.bib`.

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      simp [F.mapComp_assoc_right_inv_app, Strict.associator_eqToIso]
+
+variable (F)
+
+/-- The projection `∫ F ⥤ 𝒮` given by projecting both objects and homs to the first
+factor. -/
+@[simps]
+def forget : ∫ F ⥤ 𝒮 where
+  obj := fun X => X.base
+  map := fun f => f.base
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:121]:
```suggestion
  obj X := X.base
  map f := f.base
```

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+end
+
+/-- The category structure on `∫ F`. -/
+instance category : Category (∫ F) where
+  toCategoryStruct := Pseudofunctor.Grothendieck.categoryStruct
+  id_comp {a b} f := by
+    ext
+    · simp
+    · simp [F.mapComp_id_right_inv_app, Strict.rightUnitor_eqToIso,
+        ← (F.mapId ⟨op a.1⟩).inv.naturality_assoc f.fiber]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:102]:
```suggestion
    · simp [F.mapComp_id_right_inv_app, Strict.rightUnitor_eqToIso, ← NatTrans.naturality_assoc]
```

File: Mathlib/CategoryTheory/Bicategory/Grothendieck.lean [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  comp_id {a b} f := by
+    ext
+    · simp
+    · simp [F.mapComp_id_left_inv_app, ← Functor.map_comp_assoc, Strict.leftUnitor_eqToIso]
+  assoc f g h := by
+    ext
+    · simp
+    · dsimp
+      slice_lhs 3 4 => rw [← (F.mapComp g.base.op.toLoc f.base.op.toLoc).inv.naturality h.fiber]
+      simp [F.mapComp_assoc_right_inv_app, Strict.associator_eqToIso]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Grothendieck.lean:112]:
```suggestion
    · simp [← NatTrans.naturality_assoc, F.mapComp_assoc_right_inv_app,
        Strict.associator_eqToIso]
```

**Comment by callesonne** [comment]
Thanks a lot for the comments & golf! @joelriou

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**